rust/ql/lib/codeql/rust/controlflow/internal/CfgConsistency.qll

```diff
@@ -0,0 +1,94 @@
+/**
+ * Provides classes for recognizing control flow graph inconsistencies.
+ */
+
+private import rust
+private import codeql.rust.controlflow.internal.ControlFlowGraphImpl::Consistency as Consistency
+import Consistency
+private import codeql.rust.controlflow.ControlFlowGraph
+private import codeql.rust.controlflow.internal.ControlFlowGraphImpl as CfgImpl
+private import codeql.rust.controlflow.internal.Completion
+
+/**
+ * All `Expr` nodes are `PostOrderTree`s
+ */
+query predicate nonPostOrderExpr(Expr e, string cls) {
+  cls = e.getPrimaryQlClasses() and
+  not e instanceof LetExpr and
+  not e instanceof ParenExpr and
+  exists(AstNode last, Completion c |
+    CfgImpl::last(e, last, c) and
+    last != e and
+    c instanceof NormalCompletion
+  )
+}
+
+/**
+ * Holds if CFG scope `scope` lacks an initial AST node.  Overrides shared consistency predicate.
+ */
+query predicate scopeNoFirst(CfgScope scope) {
+  Consistency::scopeNoFirst(scope) and
+  not scope = any(Function f | not exists(f.getBody())) and
+  not scope = any(ClosureExpr c | not exists(c.getBody()))
+}
+
+/** Holds if  `be` is the `else` branch of a `let` statement that results in a panic. */
+private predicate letElsePanic(BlockExpr be) {
+  be = any(LetStmt let).getLetElse().getBlockExpr() and
+  exists(Completion c | CfgImpl::last(be, _, c) | completionIsNormal(c))
+}
+
+/**
+ * Holds if `node` is lacking a successor. Overrides shared consistency predicate.
+ */
+query predicate deadEnd(CfgImpl::Node node) {
+  Consistency::deadEnd(node) and
+  not letElsePanic(node.getAstNode())
+}
+
+/**
+ * Gets counts of control flow graph inconsistencies of each type.
+ */
+int getCfgInconsistencyCounts(string type) {
+  // total results from all the CFG consistency query predicates in:
+  //  - `codeql.rust.controlflow.internal.CfgConsistency` (this file)
+  //  - `shared.controlflow.codeql.controlflow.Cfg`
+  type = "Non-unique set representation" and
+  result = count(CfgImpl::Splits ss | Consistency::nonUniqueSetRepresentation(ss, _) | ss)
+  or
+  type = "Splitting invariant 2" and
+  result = count(AstNode n | Consistency::breakInvariant2(n, _, _, _, _, _) | n)
+  or
+  type = "Splitting invariant 3" and
+  result = count(AstNode n | Consistency::breakInvariant3(n, _, _, _, _, _) | n)
+  or
+  type = "Splitting invariant 4" and
+  result = count(AstNode n | Consistency::breakInvariant4(n, _, _, _, _, _) | n)
+  or
+  type = "Splitting invariant 5" and
+  result = count(AstNode n | Consistency::breakInvariant5(n, _, _, _, _, _) | n)
+  or
+  type = "Multiple successors of the same type" and
+  result = count(CfgNode n | Consistency::multipleSuccessors(n, _, _) | n)
+  or
+  type = "Simple and normal successors" and
+  result = count(CfgNode n | Consistency::simpleAndNormalSuccessors(n, _, _, _, _) | n)
+  or
+  type = "Dead end" and
+  result = count(CfgNode n | Consistency::deadEnd(n) | n)
```

Comment:
Am I missing something or does this use the `deadEnd` from the import on L6 and not the local `deadEnd` defined on L44? Same question for `scopeNoFirst` below.

Reply:
Thanks :)